edx/frontend-app-discussions
# [BD-03] [TNL-7411] [TNL-7412] [TNL-7413] Topics, Threads, and Posts UI

Provides React components and functionality for the Topics, Threads, and Posts interfaces. I would like to add either [MomentJS](https://momentjs.com/) or [timeago.js](https://timeago.org) in order to format datetimes for the individual Posts/Comments/Replies in a `X hours ago` format. Currently, this only directly inserts datetimes.

Please note that some reorganization has occurred:

- Posts, Comments and Replies all belong to Threads
  - There is 1 Post per Thread, and it is the very first user post in the Thread. The Post title is the Thread title
  - There are 0 or more Comments per Thread. These are the top level responses by users within the thread
  - There are 0 or more Replies per Comment. These are responses directly attributed to the parent comment
- Threads belong to Topics
  - There are 0 or more Threads per Topic
- Topics are the same as they have been in the past

**JIRA tickets**:
[BLENDED-557](https://openedx.atlassian.net/browse/BLENDED-557)
[BD-3](https://openedx.atlassian.net/browse/BD-3)
[TNL-7411](https://openedx.atlassian.net/browse/TNL-7411)
[TNL-7412](https://openedx.atlassian.net/browse/TNL-7412)
[TNL-7413](https://openedx.atlassian.net/browse/TNL-7413)

**Dependencies**: 

This PR is directly based off https://github.com/edx/frontend-app-discussions/pull/1

**Screenshots**:

### Topics UI
![topics_ui](https://user-images.githubusercontent.com/829980/90870278-910e9000-e3c3-11ea-92bb-9c835fbf6fe5.png)

### Topics UI (filtered to specific category)
![topics_subtopic_ui](https://user-images.githubusercontent.com/829980/90870276-9075f980-e3c3-11ea-9940-ee4e0c8aa94b.png)

### All Threads UI

![all_threads_ui](https://user-images.githubusercontent.com/829980/90878872-4b0bf900-e3d0-11ea-8265-cacacb4a4c94.png)

### My Threads UI

![my_threads_ui](https://user-images.githubusercontent.com/829980/90878897-52cb9d80-e3d0-11ea-8192-680555b6387a.png)

### Threads & Posts
![threads_posts_replies_ui](https://user-images.githubusercontent.com/829980/90870269-8eac3600-e3c3-11ea-9594-e94cebe873c2.png)

**Testing instructions**:

1. Go to your LMS, login as the `edx` user, and enroll in a course (any course) and create a post in the discussion forum. Repeat with at least 1 other user. Ensure one post contains replies, and there is at least 1 discussion type post, and 1 question type post.
2. Copy the course key for the course you made comments in
3. Launch the frontend-app-discussions application with `npm install && npm start`
4. Navigate to `locahost:2002/discussions/COURSE_KEY/topics`. This is the Topics UI
5. Click on the topic group (with the gray headers) containing the topic you commented in
6. Click on the topic you commented in. This is the Threads UI
7. Click on your thread. This is the Posts/Comments/Replies UI

**Author notes and concerns**:

1. I would like to add either [MomentJS](https://momentjs.com/) or [timeago.js](https://timeago.org) in order to format datetimes for the individual Posts/Comments/Replies in a `X hours ago` format. Currently, this only directly inserts datetimes.
2. There is insufficient API support for this frontend, and as such, some data/information is missing. This UI will likely need to be updated/improved slightly when the API is built/complete.

**Reviewers**
